enhancements/network/ovn-kubernetes-localnet-api.md

```diff
@@ -0,0 +1,557 @@
+| ExcludeSubnets      | List of CIDRs removed from the specified CIDRs in `subnets`.The format should match standard CIDR notation (for example, "10.128.0.0/16"). This field must be omitted if `subnetes` is unset or `ipam.mode` is `Disabled`.                                                                            | Yes      |
+| IPAM                | Contains IPAM-related configuration for the network. When ipam.lifecycle=Persistent enable workloads have persistent IP addresses. For example: Virtual Machines will have the same IP addresses along their lifecycle (stop, start migration, reboots).                                              | Yes      |
+
+#### Suggested API validations 
+- `Role`:
+  - Required.
+  - Can be Secondary only when topology=Localnet.
+    - Having Role explicitly makes the API predictable and consistent with other topologies. In addition it enables extending localnet for Primary networks.
+- `PhysicalNetworkName`: Required.
+- `MTU`:
+  - Minimum 576 (minimal for IPv4). Maximum: 65536.
+  - When Subnets consist of IPv6 CIDR, minimum MTU should be 1280.
+- `VLAN`: <br/>
+  According to [dot1q (IEEE 802.1Q)](https://ieeexplore.ieee.org/document/10004498), 
+  VID (VLAN ID) is 12-bits field, providing 4096 values; 0 - 4095. <br/>
+  The VLAN IDs `0`, `1` and `4095` are reserved. <br/> 
+  Suggested validations:
+  - Minimum: 2, Maximum: 4094.
+- `Subnets`:
+  - Minimum items 1, Maximum items 2.
+  - Items are valid CIDR (e.g.: "10.128.0.0/16")
+  - When 2 items are specified they must be of different IP families.
+- `ExcludeSubnets`:
+  - Minimum items 1, Maximum items - 25.
+  - Items are valid CIDR (e.g.: "10.128.0.0/16")
+  - Cannot be set when Subnet is unset or `ipam.mode=Disabled`.
+
+#### YAML examples
+Assuming we have defined NNCP with the following bridge mapping
+```yaml
+desiredState:
+    ovn:
+      bridge-mappings:
+      - localnet: tenantblue 
+        bridge: br-ex
+```
+Example 1:
+```yaml
+---
+apiVersion: k8s.ovn.org/v1
+kind: ClusterUserDefinedNetwork
+metadata:
+  name: test-net
+spec:
+  namespaceSelector:
+    matchExpressions:
+    - key: kubernetes.io/metadata.name
+      operator: In
+      values: ["red", "blue"]
+  network:
+    topology: Localnet
+    localnet:
+      role: Secondary
+      physicalNetworkName: tenantblue
+      subnets: ["192.168.100.0/24", "2001:dbb::/64"]
+```
+The above spec will generate the following NAD, in namespace `blue`:
+```yaml
+apiVersion: k8s.cni.cncf.io/v1
+kind: NetworkAttachmentDefinition
+metadata:
+  name: test-net
+  namespace: blue
+finalizers:
+ - k8s.ovn.org/user-defined-network-protection
+labels:
+  k8s.ovn.org/user-defined-network: ""
```

Comment:
shouldn't this label be on the namespace ?...

Reply:
The CRDs controller label NADs with this label making it easier to aggregate for maintenance by admins (view all UDN NAD, etc..).
The primary UDN label for namespace is the same.

Reply:
OK, fair enough.